proof/refine/ARM/Tcb_R.thy

```diff
@@ -2791,29 +2791,223 @@ lemma decodeSetSchedParams_wf[wp]:
+       (invs and valid_sched and (\<lambda>s. s \<turnstile> cap \<and> (\<forall>x \<in> set extras. s \<turnstile> (fst x))))
+       (invs' and (\<lambda>s. s \<turnstile>' cap' \<and> (\<forall>x \<in> set extras'. s \<turnstile>' (fst x))))
        (decode_set_sched_params args cap slot extras)
        (decodeSetSchedParams args cap' slot' extras')"
-  apply (simp add: decode_set_sched_params_def decodeSetSchedParams_def)
+  apply (simp add: decode_set_sched_params_def decodeSetSchedParams_def decode_update_sc_def
+                   check_handler_ep_def unlessE_whenE)
   apply (cases "length args < 2")
    apply (clarsimp split: list.split)
-  apply (cases "length extras < 1")
+  apply (cases "length extras < 3")
    apply (clarsimp split: list.split simp: list_all2_Cons2)
   apply (clarsimp simp: list_all2_Cons1 neq_Nil_conv val_le_length_Cons linorder_not_less)
-  sorry (*
-  apply (rule corres_split_eqrE)
-     apply (rule corres_split_norE[OF _ check_prio_corres])
-       apply (rule corres_splitEE[OF _ check_prio_corres])
-         apply (rule corres_returnOkTT)
-         apply (clarsimp simp: newroot_rel_def elim!: is_thread_cap.elims(2))
-        apply (wpsimp wp: check_prio_inv checkPrio_inv)+
-    apply (corressimp simp: valid_cap_def valid_cap'_def)+
-  done *)
+  apply (case_tac cap; clarsimp)
+  apply (rule corres_split_eqrE[THEN corres_guard_imp])
+       apply (rule corres_split_norE[OF _ check_prio_corres])
+         apply (rule corres_split_norE[OF _ check_prio_corres])
+           apply (rule corres_split_eqrE)
+              apply (clarsimp simp: liftE_bindE_bind bindE_assoc)
+              apply (rule whenE_throwError_corres; simp add: cap_rel_valid_fh)
+              apply (rule corres_returnOkTT)
+              apply (clarsimp simp: newroot_rel_def)
+             apply (rule corres_case_cap_null_sc[where Pother=\<top> and Pother'=\<top>]
+                    ; clarsimp
+                    ; (wpfix add: capability.sel)?)
+              apply (clarsimp simp: getCurThread_def)
+              apply (rule corres_split_liftEE[OF corres_gets_trivial])
+                 apply (clarsimp simp: state_relation_def)
+                apply (rule whenE_throwError_corres; clarsimp)
+                apply (rule corres_returnOkTT, simp)
+               apply wpsimp
+              apply wpsimp
+             apply (clarsimp simp: get_tcb_obj_ref_def)
+             apply (rule corres_split_eqrE)
+                apply (rule whenE_throwError_corres; clarsimp)
+                apply (rule corres_split_liftEE[OF get_sc_corres])
+                  apply (rule whenE_throwError_corres; clarsimp simp: sc_relation_def)
+                  apply (rule corres_split_liftEE[OF is_blocked_corres])
+                    apply (rule corres_split_liftEE[OF get_sc_released_corres])
+                      apply (rule whenE_throwError_corres; clarsimp)
+                      apply (rule corres_returnOkTT, simp)
+                     apply wpsimp
```

Comment:
Nitpick: here and earlier in the file, should probably combine long sequences of repeated `wpsimp` or `wpsimp wp: single_lemma` into `(wpsimp wp: all_lemmas)+`. I don't mind the style you've used, but I've been told not to do it.